### PR TITLE
fix(NODE-3071): Ignore error message if error code is defined

### PR DESCRIPTION
**Pull request by nbbeeken**
Prior behavior was to check the error message if the code did not
correspond to the server state. With this change the driver will not
inspect the error message for error type. ServerDescriptions with
topology versions less than the current will be ignored.

NODE-3071, NODE-2559

### CODE REVIEW

File: src/cmap/commands.ts
Comment (nbbeeken) [inline src/cmap/commands.ts:33]: We had an inconsistency where `socketTimeoutMS` was being ignored because we only checked for `socketTimeout`. You'll see that update in a number of places.

File: src/error.ts
Comment (nbbeeken) [inline src/error.ts:11]: Since we're using codes in a number of places it seemed to make sense to make an enum out of them all. Luckily the git diff shows the naming is the same as the comment that was placed next to a lot of the numbers.

File: src/error.ts
Comment (nbbeeken) [inline src/error.ts:334]: This additional error code is part of the spec change

File: src/error.ts
Comment (nbbeeken) [inline src/error.ts:348]: This change was to make the return type `boolean` instead of `RegExpMatch`

File: src/sdam/server_description.ts
Comment (nbbeeken) [inline src/sdam/server_description.ts:262]: When testing the counter property is a number so `===` works as expected. However the counter is supposed to be a Long and `===` as well as `<` will fail when this is the case. So converting here made sense. There might be a larger issue here of internal communications with mongod losing precision since we don't breakout our own set of bson options, thoughts?

File: src/sdam/topology.ts
Comment (nbbeeken) [inline src/sdam/topology.ts:618]: This is the changes from NODE-2559 that didn't get ported to 4.0 branch. It was necessary to do here to get spec tests to pass.

File: src/sdam/topology.ts
Comment (nbbeeken) [inline src/sdam/topology.ts:984]: Also NODE-2559

File: test/functional/spec-runner/index.js
Comment (nbbeeken) [inline test/functional/spec-runner/index.js:128]: Syncing the tests added some tests that use `authEnabled` as a runOn requirement, I believe we don't currently have a way of filtering for this, correct me if I'm wrong and I'll make the change. If not I left a comment suggesting the method the python driver uses.

File: test/functional/spec-runner/index.js
Comment (nbbeeken) [inline test/functional/spec-runner/index.js:137]: We haven't implemented the paused or ready events on a connection pool yet which some tests block on a `'waitForEvent'` operation, until mocha times them out.

File: test/functional/spec-runner/index.js
Comment (emadum) [reply test/functional/spec-runner/index.js:128]: One option could be to check the `$AUTH` environment variable. At least when running on Evergreen, it'll be `auth` if the tests are running against a server with auth enabled. A bit more hassle when running tests locally since you need to set that env var when running your tests, but the simplicity of the implementation could make that worth it.

**Comment by nbbeeken** [comment]
Hey @emadum if you can take a look when you get the chance that would be much appreciated 🙂

File: test/functional/spec-runner/index.js
Comment (emadum) [reply test/functional/spec-runner/index.js:128]: Flagging this comment because it seems I didn't leave it as a review last time.

**Review by emadum — APPROVED** [review]
Had one comment a few days ago but it looks like it got lost, either way I don't think that needs to hold this up; LGTM 👍

**Comment by nbbeeken** [comment]
Hey @durran just a reminder to take a look at this when you get the chance, thanks!

**Pull request merged**
